fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/matchers/ListChildrenMatcher.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp.matchers;
+
+import com.apple.foundationdb.record.query.plan.temp.Bindable;
+import com.apple.foundationdb.record.query.plan.temp.ExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.PlannerExpression;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A matcher for a children that tries to match each child to a specified matcher. It implements the "default" matching
+ * for children, as exposed by the {@link TypeMatcher#of(Class, ExpressionMatcher[])} helper method.
+ *
+ * <p>
+ * If every child matches the specified matcher and if every matcher matches a child, the {@link ListChildrenMatcher}
+ * returns a stream of bindings that includes the Cartesian product of the stream of bindings from each child. Note that
+ * this matcher can only match a finite number of children. Currently, matching is not especially efficient: the cross
+ * product is computed by collecting each child stream into a list.
+ * </p>
+ */
+public class ListChildrenMatcher implements ExpressionChildrenMatcher {
+    @Nonnull
+    private static final ListChildrenMatcher EMPTY = new ListChildrenMatcher(Collections.emptyList());
+
+    @Nonnull
+    private final List<ExpressionMatcher<? extends Bindable>> childMatchers;
+
+    private ListChildrenMatcher(@Nonnull List<ExpressionMatcher<? extends Bindable>> childMatchers) {
+        this.childMatchers = childMatchers;
+    }
+
+    @Nonnull
+    @Override
+    public Stream<PlannerBindings> matches(@Nonnull Iterator<? extends ExpressionRef<? extends PlannerExpression>> childIterator) {
+        Iterator<ExpressionMatcher<? extends Bindable>> bindingIterator = childMatchers.iterator();
+
+        Stream<PlannerBindings> bindingStream = Stream.of(PlannerBindings.empty());
+        while (childIterator.hasNext() && bindingIterator.hasNext()) {
+            List<PlannerBindings> possible = childIterator.next().bindTo(bindingIterator.next()).collect(Collectors.toList());
+            if (possible.isEmpty()) {
+                return Stream.empty();
+            }
+            bindingStream = bindingStream.flatMap(existing -> possible.stream().map(existing::mergeWith));
+        }
```

Comment:
Hm, I'm debating about whether the common code here and in `AllChildrenMatcher` is worth consolidating into a single "cartesian product" function. It needs to somehow solve the problem of one requiring a single matcher that is reused and the other requiring an iterator of matchers. Maybe it's not worth it.

Reply:
All right, sounds good.